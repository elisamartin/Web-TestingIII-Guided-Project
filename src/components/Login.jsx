import React from 'react';
import axios from 'axios';

export default class Login extends React.Component {
	state = { flashMessage: '', username: '', password: '' };

	onChange = (field, value) => {
		this.setState({ [field]: value });
	};

	onLogin = () => {
		axios
			.post('login', {
				username: this.state.username,
				password: this.state.password
			})
			.then((res) => {
				localStorage.setItem('token', res.data.token);
				this.setState({ flashMessage: `Welcome, ${res.data.username}` });
			})
			.catch((res) => {
				this.setState({ flashMessage: res.message });
			});
	};

	render() {
		return (
			<div>
				<div>
					<label htmlFor="usernameInput">username</label>
					<input
						id="usernameInput"
						className="input"
						onChange={(e) => this.onChange('username', e.target.value)}
						placeholder="username"
						type="text"
					/>
					<br />
					<label htmlFor="passwordInput">password</label>
					<input
						id="passwordInput"
						className="input"
						onChange={(e) => this.onChange('password', e.target.value)}
						placeholder="password"
						type="text"
					/>
				</div>
				{this.state.username &&
				this.state.password && (
					<button onClick={this.onLogin} data-testid="loginButton">
						Login
					</button>
				)}
				<div>{this.state.flashMessage}</div>
			</div>
		);
	}
}
