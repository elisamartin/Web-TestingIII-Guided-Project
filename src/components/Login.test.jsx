import React from 'react';
import * as rt from 'react-testing-library';
import Login from './Login';

afterEach(rt.cleanup);

describe('Login', () => {
	it('displays no login button if no username or no password', () => {
		const wrap = rt.render(<Login />);
		const button = wrap.queryByTestId(/loginButton/i);
		expect(button).toBeFalsy();
		expect(wrap.asFragment()).toMatchSnapshot();
	});

	// it('can change input values', () => {
	//   const wrap = rt.render(<Login />);
	//   rt.fireEvent.change(
	//     wrap.getByPlaceholderText('username'),
	//     { target: { value: 'foo' } },
	//   );

	//   rt.fireEvent.change(
	//     wrap.getByPlaceholderText('password'),
	//     { target: { value: 'bar' } },
	//   );

	//   expect(wrap.getByDisplayValue('foo'));
	//   expect(wrap.getByDisplayValue('bar'));
	// });

	it('can change input values', () => {
		const wrap = rt.render(<Login />);
		expect(wrap.asFragment()).toMatchSnapshot();

		const usernameInput = wrap.getByPlaceholderText('username');
		const passwordInput = wrap.getByPlaceholderText('password');

		const inputValue = 'name';
		const passValue = 'secretPassword';

		rt.fireEvent.change(usernameInput, { target: { value: inputValue } });
		expect(wrap.getByPlaceholderText('username').value).toBe(inputValue);

		rt.fireEvent.change(passwordInput, { target: { value: passValue } });
		expect(wrap.getByPlaceholderText('password').value).toBe(passValue);
	});

	it('displays login button if username and password', () => {
		const wrap = rt.render(<Login />);
		const usernameInput = wrap.getByLabelText('username');
		const passwordInput = wrap.getByLabelText('password');

		rt.fireEvent.change(usernameInput, { target: { value: 'A' } });

		expect(wrap.queryByTestId(/loginButton/i)).toBeFalsy();

		rt.fireEvent.change(passwordInput, { target: { value: 'B' } });

		expect(wrap.queryByTestId(/loginButton/i)).toBeTruthy();
	});

	it('can login successfully', async () => {
		// grab the component
		const wrap = rt.render(<Login />);
		// change username to Alex
		rt.fireEvent.change(wrap.getByPlaceholderText('username'), {
			target: { value: 'Alex' }
		});
		// change password to be longer than 0
		rt.fireEvent.change(wrap.getByLabelText('password'), {
			target: { value: 'secret' }
		});
		// click the button
		rt.fireEvent.click(wrap.queryByTestId(/loginButton/i));
		//await flash msg to appear
		await wrap.findByText(/welcome/i);
		// assert message
		expect(wrap.getByText(/welcome/i));
	});

	it('can fail miserably', async () => {
		// see the error render
	});
});
