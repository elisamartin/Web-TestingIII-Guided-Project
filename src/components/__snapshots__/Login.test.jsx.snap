// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login can change input values 1`] = `
<DocumentFragment>
  <div>
    <div>
      <label
        for="usernameInput"
      >
        username
      </label>
      <input
        class="input"
        id="usernameInput"
        placeholder="username"
        type="text"
      />
      <br />
      <label
        for="passwordInput"
      >
        password
      </label>
      <input
        class="input"
        id="passwordInput"
        placeholder="password"
        type="text"
      />
    </div>
    <div />
  </div>
</DocumentFragment>
`;

exports[`Login displays no login button if no username or no password 1`] = `
<DocumentFragment>
  <div>
    <div>
      <label
        for="usernameInput"
      >
        username
      </label>
      <input
        class="input"
        id="usernameInput"
        placeholder="username"
        type="text"
      />
      <br />
      <label
        for="passwordInput"
      >
        password
      </label>
      <input
        class="input"
        id="passwordInput"
        placeholder="password"
        type="text"
      />
    </div>
    <div />
  </div>
</DocumentFragment>
`;
